src/utils/daemon.rs

```diff
@@ -44,80 +45,159 @@ pub fn start_daemon() {
 
-
     // check new crates every minute
     thread::spawn(move || {
+        // space this out to prevent it from clashing against the queue-builder thread on launch
+        thread::sleep(Duration::from_secs(30));
         loop {
+            let opts = opts();
+            let mut doc_builder = DocBuilder::new(opts);
+
+            debug!("Checking new crates");
+            match doc_builder.get_new_crates() {
+                Ok(n) => debug!("{} crates added to queue", n),
+                Err(e) => error!("Failed to get new crates: {}", e),
+            }
+
             thread::sleep(Duration::from_secs(60));
+        }
+    });
+
+    // build new crates every minute
+    thread::spawn(move || {
+        let opts = opts();
+        let mut doc_builder = DocBuilder::new(opts);
+
+        /// Represents the current state of the builder thread.
+        enum BuilderState {
+            /// The builder thread has just started, and hasn't built any crates yet.
+            Fresh,
+            /// The builder has just seen an empty build queue.
+            EmptyQueue,
+            /// The builder has just seen the lock file.
+            Locked,
+            /// The builder has just finished building a crate. The enclosed count is the number of
+            /// crates built since the caches have been refreshed.
+            QueueInProgress(usize),
+        }
 
-            let mut opts = opts();
-            opts.skip_if_exists = true;
+        let mut status = BuilderState::Fresh;
+
+        loop {
+            if !status.is_in_progress() {
+                thread::sleep(Duration::from_secs(60));
+            }
 
             // check lock file
-            if opts.prefix.join("cratesfyi.lock").exists() {
+            if doc_builder.is_locked() {
                 warn!("Lock file exits, skipping building new crates");
+                status = BuilderState::Locked;
                 continue;
             }
 
-            let mut doc_builder = DocBuilder::new(opts);
+            if status.count() > 10 {
+                // periodically, we need to flush our caches and ping the hubs
+                debug!("10 builds in a row; flushing caches");
+                status = BuilderState::QueueInProgress(0);
 
-            debug!("Checking new crates");
-            let queue_count = match doc_builder.get_new_crates() {
-                Ok(size) => size,
-                Err(e) => {
-                    error!("Failed to get new crates: {}", e);
-                    continue;
+                match pubsubhubbub::ping_hubs() {
+                    Err(e) => error!("Failed to ping hub: {}", e),
+                    Ok(n) => debug!("Succesfully pinged {} hubs", n)
                 }
-            };
 
-            // Only build crates if there is any
-            if queue_count == 0 {
-                debug!("Queue is empty, going back to sleep");
-                continue;
-            }
+                if let Err(e) = doc_builder.load_cache() {
```

Comment:
This load/save dance is here in case the on-disk cache has been updated since the builder thread has last saved it. Since it's represented in-memory as a `BTreeSet`, reloading everything in doesn't create any duplicates in the cache when it's saved into disk later.